src/SIL.XForge.Identity/Controllers/Account/AccountController.cs

```diff
@@ -167,7 +176,91 @@ public async Task<IActionResult> Login(LoginInputModel model, string button)
             return View(vm);
         }
 
-        
+        /// <summary>
+        /// Show reset password page
+        /// </summary>
+        [HttpGet]
+        public async Task<IActionResult> ResetPassword(string token)
+        {
+            if (!string.IsNullOrEmpty(token))
+                return View();
+            else
+            {
+                LoginViewModel vm = await BuildLoginViewModelAsync("");
+                return View("Login", vm);
+            }
+        }
+
+        /// <summary>
+        /// Show forgot password page
+        /// </summary>
+        [HttpGet]
+        public async Task<IActionResult> ForgotPassword()
+        {
+            ForgotPasswordViewModel vm = await BuildForgotPasswordViewModelAsync();
+            return View("ForgotPassword", vm);
+        }
+
+        /// <summary>
+        /// Validate username or email and send email
+        /// </summary>
+        [HttpPost]
+        public async Task<IActionResult> ForgotPassword(ForgotPasswordViewModel model)
+        {
+            UserEntity user = await _users.Query().SingleOrDefaultAsync(u => u.Username == model.UsernameOrEmail || u.Email == model.UsernameOrEmail);
```

Comment:
You can combine the `Query` and `UpdateAsync` calls into a single `UpdateAsync` call. Use the `UpdateAsync` overload that allows you to specify a filter.

Reply:
You should be able to do something like this:
``` cs
byte[] time = BitConverter.GetBytes(DateTime.UtcNow.ToBinary()); 
byte[] key = Guid.NewGuid().ToByteArray(); 
string resetPasswordKey = Convert.ToBase64String(time.Concat(key).ToArray());
UserEntity user = await _users.UpdateAsync(u => u.Username == model.UsernameOrEmail || u.Email == model.UsernameOrEmail, update => update
    .Set(u => u.ResetPasswordKey, resetPassword)
    .Set(u => u.ResetPasswordExpirationDate, DateTime.Now.AddDays(7)));
if (user != null)
{
    // Send email and redirect to login page
}
else
{
    // Redirect to forgot password page with error message
}
````

Reply:
We implemented your logic and joined query and updateAsync together. Thank you for your logic and code.